### PR TITLE
3d Grid

### PR DESCRIPTION
**Pull request by amin1377**
In this PR we have changed DeviceGrid data structure to support multi-die FPGA architecture

### CODE REVIEW

**Comment by amin1377** [comment]
@vaughnbetz: This PR is ready for review

**Comment by amin1377** [comment]
@vaughnbetz: [Here](https://utoronto-my.sharepoint.com/:x:/g/personal/amin_mohaghegh_mail_utoronto_ca/EWqukqzqVZxFulMXoDH_dwYBphA2PsZkqLjZ1CppVqD1BA?e=vYSdxO) is the QoR results for VTR and Titan benchmark.

**Review by saaramahmoudi — COMMENTED** [review]
Just a quick reminder that we need to add documentation for device grid.

**Comment by vaughnbetz** [comment]
VTR suite placement is 1.9% slower, Titan is 6.5% slower.  That is significant.  
- Check if the results are consistent across circuits, and/or run another seed.  Is this increase real?  It probably is, since you tell me you ran on the Savi machines in a consistent way (5 threads on master, let it finish, then 5 threads on new code.
- Can we get this back?  Worth profiling etc.  We'd rather not slow down by 6% if at all possible.

File: libs/libarchfpga/src/device_grid.cpp [hunk 1]
```diff
@@ -1,41 +1,79 @@
 #include "device_grid.h"
```
Comment (vaughnbetz) [inline libs/libarchfpga/src/device_grid.cpp:2]: What is this for?  Need a comment on purpose and use.

Comment (vaughnbetz) [reply libs/libarchfpga/src/device_grid.cpp:2]: But these comments should go in the header file.

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
Just started a review to give you a flavour of what I'm looking for.

**Comment by amin1377** [comment]
> 

I ran the master branch and grid_3d for the titan benchmark again. I guess the last time I wasn't using the same number of threads, and the master branch was not the most recent one. Now the run-time of the two branches is almost similar.

**Comment by amin1377** [comment]
> Thanks, the new data looks good. I'll go through the code tonight or when we meet tomorrow and hopefully we can land this soon!

I have added comments to device_grid.h and device_grid.cpp, which are the two files where I made actual changes. The other files were modified because they previously accessed the t_grid_tile data structure directly. With the changes I made, other functions can now access this data structure only through the helper functions in DeviceGrid.

**Review by tangxifan — COMMENTED** [review]
Most of the code changes look good to me. Just need some clarification.

File: libs/libarchfpga/src/device_grid.h [hunk 2]
```diff
@@ -6,54 +6,115 @@
...
+    int num_layers_ = 1;
+
+    /**
+     * @brief grid_ is a vector of 2D grids. Each 2D grid represents a layer (die) of the FPGA chip. The layers are pushed backed to the top from the bottom (grid_[0] corresponds to the bottom layer).
+     * @note The first dimension is the layer number, the second dimension is the x coordinate, and the third dimension is the y coordinate.
+     * @note Note that vtr::Matrix operator[] returns and intermediate type
+     * @note which can be used or indexing in the second dimension, allowing
+     * @note traditional 2-d indexing to be used
+     */
+    std::vector<vtr::Matrix<t_grid_tile>> grid_;
```
[the remark this answers is not in the document]
Comment (amin1377) [reply libs/libarchfpga/src/device_grid.h:114]: The constructor of DeviceGrid would also accept a vector of 2d Matrix. VPR make the grid in "SetupGrid.cpp:build_device_grid." This function is also changes to pass a vector of grid to the DeviceGrid constructor.

File: libs/libarchfpga/src/device_grid.h [hunk 3]
```diff
@@ -6,54 +6,115 @@
...
+    ///@brief Return the width of the grid at the specified layer
+    size_t width(int layer_num = 0) const { return grid_[layer_num].dim_size(0); }
+    ///@brief Return the height of the grid at the specified layer
+    size_t height(int layer_num = 0) const { return grid_[layer_num].dim_size(1); }
 
-    //Note: supports 2-d indexing [0..width()-1][0..height()-1] yielding a t_grid_tile
-    auto operator[](size_t index) const { return grid_[index]; }
-    auto operator[](size_t index) { return grid_[index]; }
+    ///@brief Return the number of layers(number of dies)
+    inline int get_num_layers() const {
```
[the remark this answers is not in the document]
Comment (amin1377) [reply libs/libarchfpga/src/device_grid.h:36]: Agreed! I removed num_layers_.

**Comment by amin1377** [comment]
Close this PR since it is merged in modeling_multi_die_stack.

**Pull request closed without merging**